Tutorial | Custom Shape Tool in C++ and Python (#1119)
o3de/o3de.org: content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md

In this example, the signal listener uses a lambda handler, so you can define the C++ function without defining a slot.

1. Define `CreateEntityWithShapeComponent` as a standard C++ function in `ShapeExampleWidget.h`. This function accepts an `AZ::TypeId&` as a parameter so it can create an entity with the specified component type.

    ```cpp
    private:
        void CreateEntityWithShapeComponent(const AZ::TypeId& typeId);
    ```

2. In the loop that you created earlier, create a connection from the `clicked` signal in the `shapeButton` to the lambda function in `this`. The lambda function calls `CreateEntityWithShapeComponent(...)` and captures the `typeId` for the Shape component being iterated over.

    ```cpp
        QObject::connect(shapeButton, &QPushButton::clicked, this, [this, typeId]() {
            CreateEntityWithShapeComponent(typeId);
        });
    ```

### Communicate with EBuses

Define `CreateEntityWithShapeComponent(...)`, which communicates with O3DE EBuses to create a new entity with a component specified by the `typeId` parameter.

> 1. Send a request to create a new entity by calling `EditorRequestBus::BroadcastResult(...)`. Dispatch the `EditorRequests::CreateNewEntity` event, which creates a new entity and returns its `AZ::EntityId`. The new entity's `AZ::EntityId` is stored in.

Reviewer: ```suggestion
1. Send a request to create a new entity by calling `EditorRequestBus::BroadcastResult(...)`. Dispatch the `EditorRequests::CreateNewEntity` event, which creates a new entity and returns its `AZ::EntityId`. The new entity's `AZ::EntityId` is stored in `newEntityId`.
```

                AZ::ComponentTypeList{ typeId }
            );

            AZ_Assert(componentNames.size() == 1, "Unable to find component name");

            QString shapeName(componentNames[0].c_str());
            shapeName.replace(" ", "");

            entityName = QString("%1_%2").arg(entityName).arg(shapeName);
        }
        
        EditorEntityAPIBus::Event(newEntityId, &EditorEntityAPIRequests::SetName, entityName.toUtf8().constData());
    }

    // 3
    bool validFloat = false;
    float scale = m_scaleInput->currentText().toFloat(&validFloat);
    if (validFloat)
    {
        AZ::TransformBus::Event(newEntityId, &AZ::TransformInterface::SetLocalUniformScale, scale);
    }

    // 4
    EditorComponentAPIBus::Broadcast(&EditorComponentAPIRequests::AddComponentsOfType, newEntityId, AZ::ComponentTypeList{ typeId });
}
```

## Icon

An icon is an image file that's used to represent your tool in the Editor. The icon appears in the Edit Mode Toolbar in the Editor (see the following image).

{{< image-width "/images/learning-guide/tutorials/extend-the-editor/shape-example/icon.png" "500" "Add an icon for your tool in the Editor" >}}

### Add an icon

> The following instructions walk you through how to store the icon using the Qt Resource System and load it from your Gem module. Be aware that some of the instructions may already be done by the `PythonToolGem` template.

Reviewer: ```suggestion
The following instructions walk you through how to store the icon using the Qt Resource System and load it from your Gem module. Be aware that some of the instructions may already be done by the `CppToolGem` template.
```

> 1. Add an image file to the `Code/Source` directory to use as your icon. In this example, the icon is named `toolbar.svg`. We recommend that your icon adheres to the guidelines in [UI development best practices](https://o3de.org/docs/tools-ui/uidev-component-development-guidelines/#ui-development-best-practices).

Reviewer: ```suggestion
1. Add an image file to the `Code/Source` directory to use as your icon. In this example, the icon is named `toolbar.svg`. We recommend that your icon adheres to the guidelines in [UI development best practices](/docs/tools-ui/uidev-component-development-guidelines/#ui-development-best-practices).
```

> 2. Add your icon to PyShapeExample Gem's resources by updating `Code/Source/ShapeExample.qrc` with your new icon's file name.

Reviewer: ```suggestion
2. Add your icon to MyCppShapeExample Gem's resources by updating `Code/Source/MyCppShapeExample.qrc` with your new icon's file name.
```

    
    ```xml
    <!DOCTYPE RCC><RCC version="1.0">

>         <qresource prefix="/PyShapeExample">

Reviewer: ```suggestion
        <qresource prefix="/MyCppShapeExample">
```

            <file alias="toolbar_icon.svg">toolbar_icon.svg</file>
        </qresource>
    </RCC>
    ```

> 3. Register ShapeExample Gem's resources to Qt Resource System by adding the following code in `Code/Source/EditorModule.cpp`.

Reviewer: Should this be MyCppShapeExample?

>     - Define the function `InitShapeExampleResource()` and call `Q_INIT_RESOURCE(...)` to register the Qt resrouces listed in `ShapeExample.qrc`.

Reviewer: ```suggestion
    - Define the function `InitShapeExampleResources()` and call `Q_INIT_RESOURCE(...)` to register the Qt resources listed in `MyCppShapeExample.qrc`.
```

    ```cpp
    void InitShapeExampleResources()
    {

>         Q_INIT_RESOURCE(ShapeExample);

Reviewer: Should this be MyCppShapeExample?
```suggestion
        Q_INIT_RESOURCE(MyCppShapeExample);
```

    }
    ```

>    - Call `InitShapeExampleResource()` in the `EditorModule` class's constructor.

Reviewer: ```suggestion
   - Call `InitShapeExampleResources()` in the `EditorModule` class's constructor.
```

## Build and debug your tool

Build and debug your custom tool using **Visual Studio 2019**.

1. Launch Visual Studio 2019 and open the O3DE solution.

> 2. Find your Gem in the Solution Explorer. Right-click and select **Build** or **Debug**.

Reviewer: I don't think we need steps for Visual Studio, specifically. After all, they could just build again from the command line. They might also be using a different IDE for Linux. Also, there isn't really any information provided here about debugging. Maybe the title could just be named "Build and test your tool". Debugging is sort of implied.

Author: I lean towards keeping the steps in the section in because it shows how to build only the Gem, and not the entire engine, which saves time during development. At the moment, I only tested this on Windows and using Visual Studios.

For now, I can specify that these are for building with Windows, and I can change the title to "Build and test your Gem". (Changing "tool" to "Gem".)

> After, you can launch Editor and load the Gem.

Reviewer: Maybe this should be a step?
```suggestion
Open your project in the Editor.
```

Congratulations! You created a custom tool Gem that's written in C++, built it, and loaded it in the Editor. Your Shape Example tool should look something like this:

{{< image-width "/images/learning-guide/tutorials/extend-the-editor/shape-example/cpp-shape-example-ui.png" "500" "An image of the Shape Example tool." >}}

## Download the ShapeExample Gem sample

This tutorial is based off of the **ShapeExample** Gem in [`o3de/sample-code-gems` repository](https://github.com/o3de/sample-code-gems/tree/main/cpp_gems/ShapeExample). You can download this sample and load it in the Editor.

1. Download or clone the repository. The ShapeExample Gem is located in `sample-code-gems/cpp_gems/ShapeExample`.

    ```cmd
    git clone https://github.com/o3de/sample-code-gems.git
    ```

2. Before you can open the Shape Example tool, do the following:

   - Register your Gem.

   - Enable it in your project.

   - Rebuild your project.

   - Open the tool in the Editor.

>     These steps are explained earlier in this tutorial (refer to [Create a tool with the `CppToolGem` template](#create-a-tool-with-the-cpptoolgem-template)).

Reviewer: ```suggestion
    These steps are explained earlier in this tutorial. Refer to [Create a tool with the `CppToolGem` template](#create-a-gem-from-the-cpptoolgem-template).
```

2. Set the placeholder text in `m_nameInput` by calling `setPlaceholderText(...)`.

3. Enable a button to clear the text using `setClearButtonEnabled(true)`.

```cpp
    m_nameInput = new QLineEdit(this);
    m_nameInput->setPlaceholderText(QObject::tr("Set custom Entity name here..."));
    m_nameInput->setClearButtonEnabled(true);
```

### Create a checkbox

A checkbox is an option button that users can enable or disable to trigger a user-defined behavior. With Qt, you can create an checkbox by using the `QCheckBox` object. In this example, the checkbox controls whether or not to append a suffix to the entity's name. At runtime, it will start disabled, and enable when the user enters the entity's name to the input field. You will define this behavior later.

In `ShapeExampleWidget.h`, create a private `QCheckBox*` variable. In this example, name it `m_addShapeNameSuffix`.

```cpp
    private:
        //...
        QCheckBox* m_addShapeNameSuffix = nullptr;
        //...
```

In `ShapeExampleWidget.cpp`, do the following:

1. Create a checkbox by instantiating `QCheckBox` and assigning it to `m_addShapeNameSuffix`.

> 2. Set the checkbox to the start in the disabled state by using `setDisabled(true)`.

Reviewer: ```suggestion
2. Set the checkbox to start in the disabled state by using `setDisabled(true)`.
```

1. Launch Visual Studio 2019 and open the O3DE solution.

1. Find your Gem in the Solution Explorer. Right-click and select **Build** or **Debug**.

2. Open your project in the Editor and load the Gem.

Congratulations! You created a custom tool Gem that's written in C++, built it, and loaded it in the Editor. Your Shape Example tool should look something like this:

{{< image-width "/images/learning-guide/tutorials/extend-the-editor/shape-example/cpp-shape-example-ui.png" "500" "An image of the Shape Example tool." >}}

## Download the ShapeExample Gem sample

This tutorial is based off of the **ShapeExample** Gem in [`o3de/sample-code-gems` repository](https://github.com/o3de/sample-code-gems/tree/main/cpp_gems/ShapeExample). You can download this sample and load it in the Editor.

1. Download or clone the repository. The ShapeExample Gem is located in `sample-code-gems/cpp_gems/ShapeExample`.

    ```cmd
    git clone https://github.com/o3de/sample-code-gems.git
    ```

2. Before you can open the Shape Example tool, do the following:

   - Register your Gem.

   - Enable it in your project.

   - Rebuild your project.

   - Open the tool in the Editor.

>     These steps are explained earlier in this tutorial. Refer to [Create a tool with the `CppToolGem` template](#create-a-gem-from-the-cpptoolgem-template).

Reviewer: ```suggestion
    These steps are explained earlier in this tutorial. Refer to [Create a Gem from the `CppToolGem` template](#create-a-gem-from-the-cpptoolgem-template).
```
